### PR TITLE
Add a note for exposing ports while using CNI.

### PR DESCRIPTION
**Pull request by GamePlayer-8**
#### Does this PR introduce a user-facing change?

```release-note
None
```

action required

Pull request is about adding a note to docs about dealing with port exposing while using CNI networking.
Additional changes might be needed for fitting into the docs page formatting style.

### CODE REVIEW

**Comment by Luap99** [comment]
It has been a while since I used CNI with ipv6 but I do not recall ever needing to do this. Also I do not believe that there is a difference between CNI and netavark here, we use the same iptables rules.

It is not clear what you did and why this fixes it? Do you run as root or rootless? Does your cni network has ipv6 enabled?

**Comment by GamePlayer-8** [comment]
CNI has IPv6 enabled + I'm running it as root.

I though, commits says enough, but I added a note in the docs about the CNI / IPv6 problem while exposing ports.

**Comment by Luap99** [comment]
```
$ sudo podman --network-backend cni network create t1 --ipv6
t1
$ sudo podman --network-backend cni run -d -p 80:80  --network t1 nginx
892c1edf81a54e697e694f32c31617b9e5477ed65ebe9984f22885be63dd475e

$ ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: enp44s0u1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    link/ether 3c:18:a0:57:fc:9a brd ff:ff:ff:ff:ff:ff
    inet 192.168.1.134/24 brd 192.168.1.255 scope global dynamic noprefixroute enp44s0u1
       valid_lft 39582sec preferred_lft 39582sec
    inet6 fe80::7222:5dd3:d486:306d/64 scope link noprefixroute 
       valid_lft forever preferred_lft forever
5: cni-podman1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether b2:6d:1e:51:12:9c brd ff:ff:ff:ff:ff:ff
    inet 10.89.0.1/24 brd 10.89.0.255 scope global cni-podman1
       valid_lft forever preferred_lft forever
    inet6 fdf5:9337:882f:4464::1/64 scope global 
       valid_lft forever preferred_lft forever
    inet6 fe80::b06d:1eff:fe51:129c/64 scope link 
       valid_lft forever preferred_lft forever
8: veth9d191eef@if2: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master cni-podman1 state UP group default 
    link/ether 72:d4:3f:26:32:f0 brd ff:ff:ff:ff:ff:ff link-netns netns-89feb680-c8a9-dab9-f72b-e6cb5881b98e
    inet6 fe80::70d4:3fff:fe26:32f0/64 scope link 
       valid_lft forever preferred_lft forever

$ curl http://[fdf5:9337:882f:4464::1]
<!DOCTYPE html>
<html>
<head>
<title>Welcome to nginx!</title>
<style>
html { color-scheme: light dark; }
body { width: 35em; margin: 0 auto;
font-family: Tahoma, Verdana, Arial, sans-serif; }
</style>
</head>
<body>
<h1>Welcome to nginx!</h1>
<p>If you see this page, the nginx web server is successfully installed and
working. Further configuration is required.</p>

<p>For online documentation and support please refer to
<a href="http://nginx.org/">nginx.org</a>.<br/>
Commercial support is available at
<a href="http://nginx.com/">nginx.com</a>.</p>

<p><em>Thank you for using nginx.</em></p>
</body>
</html>

$ sudo ip6tables -t nat -nvL
Chain PREROUTING (policy ACCEPT 2 packets, 265 bytes)
 pkts bytes target     prot opt in     out     source               destination         
    1   122 CNI-HOSTPORT-DNAT  all      *      *       ::/0                 ::/0                 ADDRTYPE match dst-type LOCAL

Chain INPUT (policy ACCEPT 2 packets, 265 bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain OUTPUT (policy ACCEPT 2 packets, 342 bytes)
 pkts bytes target     prot opt in     out     source               destination         
    1    80 CNI-HOSTPORT-DNAT  all      *      *       ::/0                 ::/0                 ADDRTYPE match dst-type LOCAL

Chain POSTROUTING (policy ACCEPT 2 packets, 342 bytes)
 pkts bytes target     prot opt in     out     source               destination         
    3   422 CNI-HOSTPORT-MASQ  all      *      *       ::/0                 ::/0                 /* CNI portfwd requiring masquerade */
  295 25943 LIBVIRT_PRT  all      *      *       ::/0                 ::/0                
    0     0 CNI-6d98b454aa8b1a1bb67f327d  all      *      *       fdf5:9337:882f:4464::4  ::/0                 /* name: "t1" id: "892c1edf81a54e697e694f32c31617b9e5477ed65ebe9984f22885be63dd475e" */

Chain CNI-6d98b454aa8b1a1bb67f327d (1 references)
 pkts bytes target     prot opt in     out     source               destination         
    0     0 ACCEPT     all      *      *       ::/0                 fdf5:9337:882f:4464::/64  /* name: "t1" id: "892c1edf81a54e697e694f32c31617b9e5477ed65ebe9984f22885be63dd475e" */
    0     0 MASQUERADE  all      *      *       ::/0                !ff00::/8             /* name: "t1" id: "892c1edf81a54e697e694f32c31617b9e5477ed65ebe9984f22885be63dd475e" */

Chain CNI-DN-6d98b454aa8b1a1bb67f3 (1 references)
 pkts bytes target     prot opt in     out     source               destination         
    1    80 CNI-HOSTPORT-SETMARK  tcp      *      *       fdf5:9337:882f:4464::/64  ::/0                 tcp dpt:80
    1    80 DNAT       tcp      *      *       ::/0                 ::/0                 tcp dpt:80 to:[fdf5:9337:882f:4464::4]:80
...
```

This is how I tested it, it works just fine without adding `[::]` to the port forward. Please share a exact reproducer if it does not work for you.

File: docs/source/markdown/options/publish.md [hunk 1]
```diff
@@ -21,3 +21,9 @@ Host port does not have to be specified (e.g. `podman run -p 127.0.0.1::80`).
 If it is not, the container port will be randomly assigned a port on the host.
 
 Use **podman port** to see the actual mapping: `podman port $CONTAINER $CONTAINERPORT`.
+
+Note for CNI networking: 
+By default using `--publish`, `-p=[hostPort]:containerPort[/protocol]` will expose the port **only for IPv4**. 
+For IPv6 please do the following - `-p=[[::]:[hostPort]:]containerPort[/protocol]` - where `[::]` is exposing the port onto IPv6.
```
Comment (TomSweeneyRedHat) [inline docs/source/markdown/options/publish.md:27]: might be nice to have an example here.

**Comment by Luap99** [comment]
> But for some reason it doesn't want with tunneling onto IPv6 network.

I have no idea what this means? Did you try if netavark and/or rootless behaves differently?

In general I do not like to merge such a doc change because it should not be needed and if it doesn't work I rather just fix the code to make it work.

**Comment by GamePlayer-8** [comment]
Steps to reproduce:

1. Set `network_backend` to `"cni"` in `/etc/containers/containers.conf`.
2. Set `cni_plugin_dirs`:
```
cni_plugin_dirs = [
  "/usr/local/libexec/cni",
  "/usr/libexec/cni",
  "/usr/local/lib/cni",
  "/usr/lib/cni",
  "/opt/cni/bin",
]
```
3. Set `enable_ipv6` to `true`.
4. Run `podman run -it -p 8080:80 linuxserver/nginx`
5. In a second terminal run `nmap -6 ::1 -p 8080`.
6. Try again with `podman run -it -p 8080:80 -p [::]:8080:80 linuxserver/nginx`.
7. Run `nmap -6 ::1 -p 8080` again.

File: docs/source/markdown/options/publish.md (hunk 1 above)
Comment (GamePlayer-8) [reply docs/source/markdown/options/publish.md:27]: Example i.e.:
`podman run -it -p 8080:80 linuxserver/nginx` - IPv4 only.
`podman run -it -p 8080:80 -p [::]:8080:80 linuxserver/nginx` - IPv4 + IPv6.
?

Comment (TomSweeneyRedHat) [reply docs/source/markdown/options/publish.md:27]: Works for me, thx!

**Comment by Luap99** [comment]
> Steps to reproduce:
> 
>     1. Set `network_backend` to `"cni"` in `/etc/containers/containers.conf`.
> 
>     2. Set `cni_plugin_dirs`:
> 
> 
> ```
> cni_plugin_dirs = [
>   "/usr/local/libexec/cni",
>   "/usr/libexec/cni",
>   "/usr/local/lib/cni",
>   "/usr/lib/cni",
>   "/opt/cni/bin",
> ]
> ```
> 
>     3. Set `enable_ipv6` to `true`.
> 
>     4. Run `podman run -it -p 8080:80 linuxserver/nginx`
> 
>     5. In a second terminal run `nmap -6 ::1 -p 8080`.
> 
>     6. Try again with `podman run -it -p 8080:80 -p [::]:8080:80 linuxserver/nginx`.
> 
>     7. Run `nmap -6 ::1 -p 8080` again.

Note that port forwarding via `::1` is not supported when you run as root via cni/netavark. The kernel does not support DNAT on `::1` so we cannot add port forward rules. You should check for actual connection traffic not just nmap to see if it is working. Podman will always bind the addresses to make sure no other process is trying to use them. I think what you are seeing here is that we only bind `0.0.0.0` by default and not `::`, this is likely something that needs to be fixed and not the docs.

**Comment by GamePlayer-8** [comment]
> I think what you are seeing here is that we only bind `0.0.0.0` by default and not `::`, this is likely something that needs to be fixed and not the docs.

I agree. So close the PR?

**Comment by Luap99** [comment]
Lets close it, could you create a new issue with that information so we/I do not forget it ?

**Pull request closed without merging**